scalyr_agent/tests/monitors_manager_test.py

```diff
@@ -74,6 +83,133 @@ def test_module_with_id(self):
+        1 monitor returns non-empty frags. 1 returns None.
+        We then verify that the callback is invoked correctly, only if the fragments change.
+
+        Notes:
+        - FakeClocks are used to avoid having to wait for thread loops.
+        - A FakeAgentLogger is defined so we can modify local test state
+        - Even though this test could have been broken up into 2 smaller ones, it is kept as one to minimize overhead
+            time taken by stop_manager()
+        """
+        counter = Counter()
+        counter_lock = threading.Lock()
+        test_frag = 'some_frag'
+
+        # Create MonitorsManager + 2 monitors. Set each to daemon otherwise unit test doesn't terminate
+        test_manager = self.__create_test_instance([
+            {
+                'module': 'scalyr_agent.builtin_monitors.test_monitor',
+                'gauss_mean': 0,
+            },
+            {
+                'module': 'scalyr_agent.builtin_monitors.test_monitor',
+                'gauss_mean': 0,
+            },
+            {
+                'module': 'scalyr_agent.builtin_monitors.test_monitor',
+                'gauss_mean': 0,
+            },
+        ], [])
+        self.assertEquals(len(test_manager.monitors), 3)
+        patched_monitor_0 = test_manager.monitors[0]
+        patched_monitor_1 = test_manager.monitors[1]
+        unpatched_monitor = test_manager.monitors[2]
+
+        # Mock the function that returns user_agent_fragment (normally invoked on a Monitor)
+        def mock_get_user_agent_fragment():
+            # 2 patched monitors will call this function concurrently
+            counter_lock.acquire()
+            try:
+                counter['fragment_polls'] += 1
+            finally:
+                counter_lock.release();
+            return test_frag
+        for mon in [patched_monitor_0, patched_monitor_1]:
+            mon.get_user_agent_fragment = mock_get_user_agent_fragment
+            self.assertEquals(mon.get_user_agent_fragment(), test_frag)  # monkey patched
+        self.assertEquals(unpatched_monitor.get_user_agent_fragment(), None)  # not patched
+
+        # Mock the callback (that would normally be invoked on ScalyrClientSession)
+        # Check that the exact fragment returned by the 2 patched monitors are deduplicated.
+        def augment_user_agent(fragments):
+            counter['callback_invocations'] += 1
+            self.assertEquals(fragments, [test_frag])
+        test_manager.set_user_agent_augment_callback(augment_user_agent)
+
+        # Fake the clock to fast-forward MonitorsManager sleep loop
+        fake_clock = scalyr_util.FakeClock()
+        for obj in [test_manager, patched_monitor_0, unpatched_monitor]:
+            obj._run_state = scalyr_util.RunState(fake_clock=fake_clock)
+
+        # Override Agent Logger to increment counter
+        # We will use this as a simple synchronization barrier to know when the monitors have finally started emitting
```

Comment:
I'll suggest an alternative approach here for this multi-thread testing.  I would argue that we don't really care about when the monitors start logging.  What we really care about is when we start getting `fragment_polls`.  So, what we should do it just block until we start getting those.

Basically, create a small thread-safe object that stores how many polls have happened.  Give that object a way of saying `block_until_n_polls` where you specify the number of polls and then a timeout (real-time) of how long you are willing to wait for that happen.  You want the timeout in case the test is broken..  you don't want to be waiting forever.

So, you start the test and block until you get the first poll.  Then, you can also do that loop you have done below where you keep incrementing the fake time.. you just block for n+1 pulls.  If you always specify a real-time timeout, you also don't have to have your check for the real clock.  If the test fails if the poll doesn't happen in 5 seconds of real time.. then it will definitely fail if it is not using a fake clock (because it takes 60 seconds between polls.)